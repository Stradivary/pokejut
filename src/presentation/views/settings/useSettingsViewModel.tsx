import { useSimulator } from "@/domain/use-cases/simulator";
import { useMantineColorScheme } from "@mantine/core";
import { modals } from "@mantine/modals";
import { notifications } from "@mantine/notifications";
import { useCallback, useMemo } from "react";

export function useSettingsViewModel() {
  
  const colorSchemeOptions = [
    {
      value: "light",
      label: "Terang",
    },
    {
      value: "dark",
      label: "Gelap",
    },
  ];

  const { setColorScheme, colorScheme } = useMantineColorScheme();


  const { pokemonList, clearPokemonList, clearSelectedPokemon } =
    useSimulator();

  const canReleaseCollection = useMemo(
    () => pokemonList.length === 0,
    [pokemonList]
  );


  const handleReleaseCollection = useCallback(
    () =>
      onReleaseCollection(
        clearSelectedPokemon,
        clearPokemonList,
        notifications
      ),
    [clearPokemonList, clearSelectedPokemon]
  );


  return {
    colorScheme,
    setColorScheme,
    colorSchemeOptions,
    handleReleaseCollection,
    canReleaseCollection,
  };
}

export function onClearCache(setRerender, navigate) {
  localStorage.clear();
  notifications.show({
    title: "Berhasil",
    message: "Cache berhasil dihapus",
    color: "blue",
    icon: <img src="/pokeball.png" alt="pokeball" />,
  });
  setRerender((e) => e + 1);
  navigate("/settings", { replace: true });
}

export function onReleaseCollection(
  clearSelectedPokemon: () => void,
  clearPokemonList: () => void,
  notifications: any
) {
  modals.openConfirmModal({
    title: "Lepas Semua Pokemon",
    children: "Apakah kamu yakin ingin melepaskan semua koleksi pokemon?",
    onConfirm: () =>
      confirmReleaseCollection(
        clearSelectedPokemon,
        clearPokemonList,
        notifications
      ),
    labels: {
      confirm: "Lepas Semua",
      cancel: "Batal",
    },
  });
}

export function confirmReleaseCollection(
  clearSelectedPokemon: () => void,
  clearPokemonList: () => void,
  notifications: any
): void {
  clearSelectedPokemon();
  clearPokemonList();
  notifications.show({
    title: "Berhasil",
    message: "Koleksi Pokemon berhasil dihapus",
    color: "blue",
    icon: <img src="/pokeball.png" alt="pokeball" />,
  });
}
