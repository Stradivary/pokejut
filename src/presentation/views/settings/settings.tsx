import { Button, Divider, Group, SegmentedControl, Stack, Title } from "@mantine/core";
import { useSettingsViewModel } from "./useSettingsViewModel";

export function Component() {
    const bindings = useSettingsViewModel();

    return <Stack p={16}>
        <Title order={1} fz={18} >Pengaturan</Title>
        <Group w="80%" mt={16} justify="space-between">
            <Title order={2} fz={16} fw="bold">Tema</Title>
            <SegmentedControl
                value={bindings.colorScheme}
                onChange={bindings.handleChangeColorScheme}
                data={bindings.colorSchemeOptions} fullWidth />
        </Group>
        <Divider />

        <Title order={2} fz={16} fw="bold">Koleksi Pokemon</Title>

        <Button
            maw={240}
            disabled={bindings.canReleaseCollection}
            color="red"
            onClick={bindings.handleReleaseCollection}>
            Hapus Koleksi Pokemon
        </Button>

    </Stack>;
}

Component.displayName = 'SettingsPage';