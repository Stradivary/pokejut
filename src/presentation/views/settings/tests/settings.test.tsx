import { beforeAll, describe, expect, it, vi } from "vitest";
import { render } from "~/tests/test-utils";
import { Component as Settings } from "../settings";
import snapshotSanitizer from "~/tests/snapshot-sanitizer";



vi.mock('../useSettingsViewModel', () => ({
  useSettingsViewModel: vi.fn(() => ({
    colorScheme: "light",
    setColorScheme: vi.fn(),
    colorSchemeOptions: ["light", "dark"],
    cacheSizeInMB: 10,
    handleClearCache: vi.fn(),
    canReleaseCollection: false,
    handleReleaseCollection: vi.fn(),
  }))
}));

describe("Settings Component", () => {
  beforeAll(() => {
    expect.addSnapshotSerializer(snapshotSanitizer);
  });
  it("should render the component correctly", () => {
    const { container } = render(<Settings />);
    expect(container).toMatchSnapshot();
  });
});

