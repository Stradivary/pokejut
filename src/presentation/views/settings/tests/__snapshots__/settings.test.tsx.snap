// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Settings Component > should render the component correctly 1`] = `
<div>
  <style
    data-mantine-styles="true"
  >
    :root{--mantine-font-family: Poppins, sans-serif;--mantine-font-family-headings: Poppins, sans-serif;--mantine-radius-default: calc(0.5rem * var(--mantine-scale));}
  </style>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 35.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 36em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 61.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 62em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 87.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 88em) {.mantine-hidden-from-xl {display: none !important;}}
  </style>
  <div
    class="mantine-Stack-root"
    style="--stack-gap: var(--mantine-spacing-md); --stack-align: stretch; --stack-justify: flex-start;"
  >
    <h1
      class="mantine-Title-root"
      data-order="1"
      style="--title-fw: var(--mantine-h1-font-weight); --title-lh: var(--mantine-h1-line-height); --title-fz: var(--mantine-h1-font-size);"
    >
      Pengaturan
    </h1>
    <div
      class="mantine-Group-root"
      style="--group-gap: var(--mantine-spacing-md); --group-align: center; --group-justify: space-between; --group-wrap: wrap; width: 80%;"
    >
      <h4
        class="mantine-Title-root"
        data-order="4"
        style="--title-fw: var(--mantine-h4-font-weight); --title-lh: var(--mantine-h4-line-height); --title-fz: var(--mantine-h4-font-size); font-weight: bold;"
      >
        Tema
      </h4>
      <div
        class="mantine-SegmentedControl-root"
        data-full-width="true"
        data-initialization="true"
        role="radiogroup"
        style="--sc-shadow: var(--mantine-shadow-xs);"
      >
        <span
          class="mantine-SegmentedControl-indicator"
          style="--sc-indicator-width: 0px; --sc-indicator-height: 0px; --sc-indicator-transform: translate(-4px, -4px);"
        />
        <div
          class="mantine-SegmentedControl-control"
          data-active="true"
          style=""
        >
          <input
            class="mantine-SegmentedControl-input"
            data-focus-ring="auto"
            id=""
            name=""
            style=""
            type="radio"
            value="light"
          />
          <label
            class="mantine-SegmentedControl-label"
            data-active="true"
            for=""
            style=""
          >
            light
          </label>
        </div>
        <div
          class="mantine-SegmentedControl-control"
          style=""
        >
          <input
            class="mantine-SegmentedControl-input"
            data-focus-ring="auto"
            id=""
            name=""
            style=""
            type="radio"
            value="dark"
          />
          <label
            class="mantine-SegmentedControl-label"
            for=""
            style=""
          >
            dark
          </label>
        </div>
      </div>
    </div>
    <div
      class="mantine-Divider-root"
      data-orientation="horizontal"
      role="separator"
      style=""
    />
    <h4
      class="mantine-Title-root"
      data-order="4"
      style="--title-fw: var(--mantine-h4-font-weight); --title-lh: var(--mantine-h4-line-height); --title-fz: var(--mantine-h4-font-size); font-weight: bold;"
    >
      Cache
    </h4>
    <p
      class="mantine-focus-auto mantine-Text-root"
      style=""
    >
      Ukuran : 
      10
       MB
    </p>
    <button
      class="mantine-focus-auto mantine-active mantine-Button-root mantine-UnstyledButton-root"
      style="--button-bg: var(--mantine-color-red-filled); --button-hover: var(--mantine-color-red-filled-hover); --button-color: var(--mantine-color-white); --button-bd: calc(0.0625rem * var(--mantine-scale)) solid transparent; max-width: calc(15rem * var(--mantine-scale));"
      type="button"
    >
      <span
        aria-hidden="true"
        class="mantine-Button-loader"
        style=""
      >
        <span
          class="mantine-Loader-root"
          style="--loader-size: calc(var(--button-height) / 1.8); --loader-color: var(--button-color);"
        />
      </span>
      <span
        class="mantine-Button-inner"
        style=""
      >
        <span
          class="mantine-Button-label"
          style=""
        >
          Hapus Cache
        </span>
      </span>
    </button>
    <div
      class="mantine-Divider-root"
      data-orientation="horizontal"
      role="separator"
      style=""
    />
    <h4
      class="mantine-Title-root"
      data-order="4"
      style="--title-fw: var(--mantine-h4-font-weight); --title-lh: var(--mantine-h4-line-height); --title-fz: var(--mantine-h4-font-size); font-weight: bold;"
    >
      Koleksi Pokemon
    </h4>
    <button
      class="mantine-focus-auto mantine-active mantine-Button-root mantine-UnstyledButton-root"
      style="--button-bg: var(--mantine-color-red-filled); --button-hover: var(--mantine-color-red-filled-hover); --button-color: var(--mantine-color-white); --button-bd: calc(0.0625rem * var(--mantine-scale)) solid transparent; max-width: calc(15rem * var(--mantine-scale));"
      type="button"
    >
      <span
        aria-hidden="true"
        class="mantine-Button-loader"
        style=""
      >
        <span
          class="mantine-Loader-root"
          style="--loader-size: calc(var(--button-height) / 1.8); --loader-color: var(--button-color);"
        />
      </span>
      <span
        class="mantine-Button-inner"
        style=""
      >
        <span
          class="mantine-Button-label"
          style=""
        >
          Hapus Koleksi Pokemon
        </span>
      </span>
    </button>
  </div>
</div>
`;
